### DISCUSSION TITLE
Cake.Frosting: How to run a task even though a previous task failed?

### DISCUSSION THREAD

**Discussion by StefanOssendorf** [cake-build/cake#3809] [Q&A] [answered]
Hello,

I've migrated our cake script to cake frosting and merged some of the azure pipelines into the cake pipeline.
Now I'm facing a problem with "publish test results".

My tasks are structured roughly:
A -> B -> C -> Dotnet Test -> Publish Test results
But when "Dotnet Test" fails the "Publish Test results" task is not executed. Is there a way to say "Execute even if one of the previous tasks failed"?

Thanks in advance
Stefan

**Comment by nils-a** [comment, marked as answer]
@StefanOssendorf kind of. You can override `OnError` in the task that fails (not the following) and make Cake not abort the build on that error.

As an example:

```cs
using System;
using Cake.Common.Diagnostics;
using Cake.Core;
using Cake.Frosting;

public static class Program
{
    public static int Main(string[] args)
    {
        return new CakeHost()
            .Run(args);
    }
}

[TaskName("WillFail")]
public sealed class WillFailTask : FrostingTask
{
    public override void Run(ICakeContext context)
    {
        throw new Exception("Boom!");
    }

    public override void OnError(Exception exception, ICakeContext context)
    {
        context.Information($"Well, well, an error of {exception.GetType().Name} occured.");
    }
}

[TaskName("Default")]
[IsDependentOn(typeof(WillFailTask))]
public class DefaultTask : FrostingTask
{
    public override void Run(ICakeContext context)
    {
        context.Information("Hello, from Default!");
    }
}
```

**Reply by StefanOssendorf** [reply]
Hi. That would make it possible to execute the next steps, yes. But it would make the whole pipeline succeed even though one step has failed.
So I need to track any failure and a "fail safe" task at the end of each pipeline (or even global teardown) and make it fail if any step failed but I needed them to continue. 🤔

**Reply by nils-a** [reply]
So, you want the tasks to continue as if no exception occurred, but then fail the whole build in the end?

**Reply by StefanOssendorf** [reply]
To your question: yes.

I want to be able to configure a task to be run even if a previous task has failed to provide information. In my case uploading the test results for the azure pipeline "test result" view but still fail the whole pipeline since a test failed.

Similiar to the feature of the azure pipeline classic editor.
![grafik](https://user-images.githubusercontent.com/5385401/153600884-0a219cdc-783d-4108-a947-21dadc32b844.png)

**Reply by nils-a** [reply]
I *think* this is not possible out-of-the-box. Maybe someone else has a better idea than I do.
My go at this would be:

```cs
using System;
using System.Collections.Generic;
using System.Linq;
using Cake.Common.Diagnostics;
using Cake.Core;
using Cake.Frosting;

public static class Program
{
    public static int Main(string[] args)
    {
        return new CakeHost()
            .UseContext<MyContext>()
            .UseTeardown<TeardownBuild>()
            .Run(args);
    }
}

public sealed class MyContext : FrostingContext
{
    public List<Exception> Exceptions = new ();
    
    public MyContext(ICakeContext context) : base(context)
    {
    }
}

public sealed class TeardownBuild : FrostingTeardown<MyContext>
{
    public override void Teardown(MyContext context, ITeardownContext info)
    {
        var lastEx = context.Exceptions.LastOrDefault();
        if (lastEx != null)
        {
            context.Error("There were exceptions in one of the previous tasks.");
            throw lastEx;
        }
    }
}

[TaskName("WillFail")]
public sealed class WillFailTask : FrostingTask<MyContext>
{
    public override void Run(MyContext context)
    {
        throw new Exception("Boom!");
    }

    public override void OnError(Exception exception, MyContext context)
    {
        context.Exceptions.Add(exception);
        context.Information($"Well, well, an error of {exception.GetType().Name} occured.");
    }
}

[TaskName("Default")]
[IsDependentOn(typeof(WillFailTask))]
public class DefaultTask : FrostingTask
{
    public override void Run(ICakeContext context)
    {
        context.Information("Hello, from Default!");
    }
}
```

Additionally, you could override `ShouldRun` in the tasks that can be skipped if an error occurred previously.

**Reply by StefanOssendorf** [reply]
Thank you for the example. I think I'll try that way next week. Maybe someone else answers in the meantime with a built-in answer :-).
Otherwise I'll mark this as answered next week.
Thanks again and have a nice weekend.

**Reply by devlead** [reply]
One way to do this would be to set a flag on the Context that would indicate a fail and fail in Teardown.

If it's just the test results you always want, then you could upload them in a try-finally block.
```cake
try
{
...DotNetTest...
}
finally
{ 
  AzurePipelines.Commands
        .PublishTestResults(new AzurePipelinesPublishTestResultsData {
            TestResultsFiles = testResults,
            MergeTestResults = true,
            Configuration = configuration,
            TestRunTitle = runTitle,
            TestRunner = AzurePipelinesTestRunnerType.VSTest
        });
}
```

https://cakebuild.net/api/Cake.Common.Build.AzurePipelines/IAzurePipelinesCommands/31048D96

**Reply by nils-a** [reply]
... or override `Finally` in the `FrostingTask`.

**Reply by StefanOssendorf** [reply]
I use a flag to fail in the global tear down. Thanks for your answers :-)
